lahja/endpoint.py

```diff
@@ -341,10 +381,24 @@ def broadcast(self, item: BaseEvent, config: Optional[BroadcastConfig] = None) -
             self._internal_queue.put_nowait((item, config))
         else:
             # Broadcast to every connected Endpoint that is allowed to receive the event
-            for name, connector in self._connected_endpoints.items():
-                allowed = (config is None) or config.allowed_to_receive(name)
-                if allowed:
-                    connector.put_nowait((item, config))
+            for name, listening_endpoint in self._listening_endpoints.items():
+
+                if config is None or config.filter_endpoint is None:
+                    # Check if a `filter_predicate` applies
+                    if listening_endpoint.config.filter_predicate is None:
```

Comment:
Up-front I want to say i'm not sure this suggestion is actually better but I want to toss it out and see if it sticks.

This value checking seems.... ?unfortunate?.  Lots of logic branches that *might* be reducable if the `None` version of at least one of these was able to be captured via the same API as when it is present.  I.E. maybe there should **always** be a `filter_predicate` that unconditionally returns `True`.

Similarly with the `config`, could there be a default *empty* config which gets attached by default which exhibits the same behavior as if it was left out?

This article is topical to this.  https://blog.conjur.org/special-cases-are-a-code-smell/

Reply:
I agree. In fact, a lot of this is based on a misconception that I had around Pythons default class attribute handling that scared me. I thought that the following was a footgun in Python

```python
class Foo:
    y = ''
``` 
Similar to
```python
class Foo:
    y = []
``` 

But while the latter is indeed a footgun, the former isn't because strings are immutable in Python. Based on this wrong assumption I used a lot of `Optiona[str]` what could have been just `str` with strings preassigned to `''`.

Will clean it up.